Sources/Main/Services/FormPost.swift

```diff
@@ -16,47 +16,94 @@
 import Foundation
 import SwiftyJSON
 
-/// A struct representing a form POST request.
+/// An implementation of form POST ``Request``.
 public struct FormPost: Request {
- public typealias Response = AuthorizationRequest
-
- /// The HTTP method for the request.
- public var method: HTTPMethod { .POST }
-
- /// Additional headers to include in the request.
- public var additionalHeaders: [String: String] = [:]
-
- /// The URL for the request.
- public var url: URL
-
- /// The request body as data.
- public var body: Data? {
-   var formDataComponents = URLComponents()
-   formDataComponents.queryItems = formData.toQueryItems()
-   let formDataString = formDataComponents.query
-   
-   if let formDataString, formDataString.isEmpty {
-     return JSON(formData).rawString()?.data(using: .utf8)
-   } else {
-     return formDataString?.data(using: .utf8)
-   }
- }
-
- /// The form data for the request.
- let formData: [String: Any]
-
- /// The URL request representation of the DirectPost.
- var urlRequest: URLRequest {
-   var request = URLRequest(url: url)
-   request.httpMethod = method.rawValue
-   request.httpBody = body
-   
-   // request.allHTTPHeaderFields = additionalHeaders
-   for (key, value) in additionalHeaders {
-     request.addValue(value, forHTTPHeaderField: key)
-   }
-   
-   return request
- }
+  public typealias Response = AuthorizationRequest
+
+  // MARK: - Request
+  public var method: HTTPMethod { .POST }
+  public let url: URL
+  public let additionalHeaders: [String: String]
+  public let body: Data?
+
+  /// The URL request representation of the DirectPost.
+  var urlRequest: URLRequest {
+    var request = URLRequest(url: url)
+    request.httpMethod = method.rawValue
+    request.httpBody = body
+
+    // request.allHTTPHeaderFields = additionalHeaders
+    for (key, value) in additionalHeaders {
+      request.addValue(value, forHTTPHeaderField: key)
+    }
+
+    return request
+  }
+
+  init(
+    url: URL,
+    /// The content type of the request body.
+    contentType: ContentType,
+    additionalHeaders: [String: String] = [:],
+    /// The form data for the request body.
+    formData: [String: Any]
+  ) {
+    self.additionalHeaders = [
+      ContentType.key: contentType.rawValue
+    ].merging(additionalHeaders) { (_, new) in new }
+    self.url = url
+    switch contentType {
+    case .form:
+      self.body = FormURLEncoded.body(from: formData)
+    case .json:
+      self.body = try? JSON(formData).rawData()
```

Comment:
Should the initializer be declared `throws` as now encoding errors are silently ignored and result in an empty (`nil`) `body`?

Reply:
Changed the initializer to throwing in b88c783.